src/components/walletconnect/HeaderWidget/index.tsx

```diff
@@ -25,12 +28,92 @@ const usePrepopulatedUri = (): [string, () => void] => {
   return [uri, clearUri]
 }
 
+const BANNER_TIMEOUT = 2_000
+
+const useSuccessMetadata = (
+  onCloseSessionManager: () => void,
+): [CoreTypes.Metadata | undefined, Dispatch<SetStateAction<CoreTypes.Metadata | undefined>>] => {
+  const { walletConnect } = useContext(WalletConnectContext)
+  const [metadata, setMetadata] = useState<CoreTypes.Metadata>()
+
+  const onSuccess = useCallback(
+    ({ peer }: SessionTypes.Struct) => {
+      onCloseSessionManager()
+
+      // Show success banner
+      setMetadata(peer.metadata)
+
+      return setTimeout(() => {
+        setMetadata(undefined)
+      }, BANNER_TIMEOUT)
+    },
+    [onCloseSessionManager],
+  )
+
+  useEffect(() => {
+    if (!walletConnect) return
+
+    let timeout: NodeJS.Timeout
+
+    walletConnect.onSessionAdd((session) => {
+      timeout = onSuccess(session)
+    })
+
+    return () => clearTimeout(timeout)
+  }, [onSuccess, walletConnect])
+
+  return [metadata, setMetadata]
+}
+
+const useDeleteMetadata = (): [
+  CoreTypes.Metadata | undefined,
+  Dispatch<SetStateAction<CoreTypes.Metadata | undefined>>,
+] => {
+  const { walletConnect } = useContext(WalletConnectContext)
+  const { safe } = useSafeInfo()
+  const [metadata, setMetadata] = useState<CoreTypes.Metadata>()
+
+  const onDelete = useCallback(
+    ({ optionalNamespaces, requiredNamespaces, peer }: SessionTypes.Struct) => {
+      const supportedEip155ChainIds = getSupportedEip155ChainIds(requiredNamespaces, optionalNamespaces)
+
+      const eipChainId = getEip155ChainId(safe.chainId)
+      const isUnsupportedChain = !supportedEip155ChainIds.includes(eipChainId)
+
+      if (!isUnsupportedChain) {
+        return
+      }
+
+      // Show success banner
+      setMetadata(peer.metadata)
+
+      return setTimeout(() => {
+        setMetadata(undefined)
+      }, BANNER_TIMEOUT * 2)
+    },
+    [safe.chainId],
+  )
+
+  useEffect(() => {
+    if (!walletConnect) return
+
+    let timeout: NodeJS.Timeout | undefined
+
+    walletConnect.onSessionDelete((session) => {
+      timeout = onDelete(session)
+    })
+
+    return () => clearTimeout(timeout)
+  }, [onDelete, walletConnect])
+
+  return [metadata, setMetadata]
+}
```

Comment:
There's a lot going on here. Please revisit this code. I'm not going to point out what exactly is not ideal in terms of code structure, I hope you can see it yourself.

Reply:
I've refactored the above and created a `useDeferredListener` hook in 231ac0c90f6dd4c8c1e650b19d55932e667df6f1. Lmk what you think.